src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -67,12 +67,20 @@ public static int LeadingZeroCount(uint value)
                 return ArmBase.LeadingZeroCount(value);
             }
 
-            // Unguarded fallback contract is 0->31
+            // Unguarded fallback contract is 0->31, BSR contract is 0->undefined
             if (value == 0)
             {
                 return 32;
             }
 
+            if (X86Base.IsSupported)
+            {
+                // LZCNT returns index starting from MSB, whereas BSR gives the index from LSB.
+                // The XOR here is equivalent to subtraction since the BSR result is always between 0 and 31.
+                // This saves one instruction, as subtraction from constant requires either MOV/SUB or NEG/ADD.
+                return 31 ^ (int)X86Base.BitScanReverse(value);
+            }
+
             return 31 - Log2SoftwareFallback(value);
```

Comment:
Do the xor trick here as well?

Reply:
Sure thing.  Also could apply to the ARM versions, but I don't know enough about their codegen to know if it makes a difference there.  Certainly couldn't hurt.